{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "from copy import deepcopy\n",
    "import time, os, sys\n",
    "\n",
    "from sklearn import linear_model, metrics\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/media/eys/xwj/mywork_B-GEX/\")\n",
    "gene_tpm_raw = pd.read_csv(\"GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.no_pseudogene\", sep=\"\\t\", skiprows=2)\n",
    "\n",
    "samples = pd.read_csv(\"GTEx_v7_Annotations_SampleAttributesDS.txt\", sep=\"\\t\")\n",
    "samples = samples[samples.SMAFRZE == 'RNASEQ']\n",
    "\n",
    "'''add one column-SUBJID '''\n",
    "samples.insert(loc=0, column='SUBJID', value =np.repeat(0,samples.shape[0]))\n",
    "'''SUBJID is made from SAMPID'''\n",
    "samples=samples.reset_index(drop=True)\n",
    "for i in range(samples.shape[0]):\n",
    "    temp=samples.loc[i,'SAMPID'].split('-')\n",
    "    samples.loc[i,'SUBJID']=temp[0]+'-'+temp[1]\n",
    "del temp\n",
    "\n",
    "'''blood and tissue RNAseq samples basic information'''\n",
    "TISSUE_list = samples.loc[(samples.SMTSD != 'Whole Blood'), ['SMTS','SMTSD']]\n",
    "TISSUE_list = TISSUE_list.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "feature_id = samples[(samples.SMTSD == 'Whole Blood') & (samples.SMAFRZE == 'RNASEQ')].loc[:,['SUBJID','SAMPID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_tpm_raw[\"Name\"]=[x.split(\".\")[0] for x in gene_tpm_raw[\"Name\"]] \n",
    "\n",
    "gene_tpm= deepcopy(gene_tpm_raw)\n",
    "gene_tpm.iloc[:, 2:] = np.log(gene_tpm.iloc[:, 2:]+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>Bsample</th>\n",
       "      <th>Tsample</th>\n",
       "      <th>Bgenes</th>\n",
       "      <th>Tgenes</th>\n",
       "      <th>BTsample</th>\n",
       "      <th>TgenesClean</th>\n",
       "      <th>BgenesClean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>407.0</td>\n",
       "      <td>442.0</td>\n",
       "      <td>42075.0</td>\n",
       "      <td>42075.0</td>\n",
       "      <td>278.0</td>\n",
       "      <td>15917.0</td>\n",
       "      <td>12121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>407.0</td>\n",
       "      <td>564.0</td>\n",
       "      <td>42075.0</td>\n",
       "      <td>42075.0</td>\n",
       "      <td>341.0</td>\n",
       "      <td>14006.0</td>\n",
       "      <td>12000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             SMTS                   SMTSD  Bsample  Tsample   Bgenes   Tgenes  \\\n",
       "0  Adipose Tissue  Adipose - Subcutaneous    407.0    442.0  42075.0  42075.0   \n",
       "1          Muscle       Muscle - Skeletal    407.0    564.0  42075.0  42075.0   \n",
       "\n",
       "   BTsample  TgenesClean  BgenesClean  \n",
       "0     278.0      15917.0      12121.0  \n",
       "1     341.0      14006.0      12000.0  "
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for t in TISSUE_list.index:\n",
    "    TISSUE = TISSUE_list.SMTSD[t]\n",
    "    target_id = samples[(samples.SMTSD == TISSUE ) & (samples.SMAFRZE == 'RNASEQ')].loc[:,['SUBJID','SAMPID']]\n",
    "    sample_overlap = list(set(target_id.SUBJID) & set(feature_id.SUBJID))\n",
    "    feature_tpm = gene_tpm[[\"Name\", \"Description\"] + list(feature_id[feature_id.SUBJID.isin(sample_overlap)].SAMPID)]\n",
    "    target_tpm = gene_tpm[[\"Name\", \"Description\"] + list(target_id[target_id.SUBJID.isin(sample_overlap)].SAMPID)]\n",
    "    # fill tissue sample number statistics\n",
    "    TISSUE_list.loc[t,'Bsample'] = feature_id.shape[0]\n",
    "    TISSUE_list.loc[t,'Tsample'] = target_id.shape[0]\n",
    "    TISSUE_list.loc[t,'Bgenes'] = feature_tpm.shape[0]\n",
    "    TISSUE_list.loc[t,'Tgenes'] = target_tpm.shape[0]\n",
    "    TISSUE_list.loc[t,'BTsample'] = len(sample_overlap)\n",
    "    TISSUE_list.loc[t,'TgenesClean'] = target_tpm.loc[target_tpm.all(axis=1)].shape[0]\n",
    "    TISSUE_list.loc[t,'BgenesClean'] = feature_tpm.loc[feature_tpm.all(axis=1)].shape[0]\n",
    "\n",
    "TISSUE_list_short = TISSUE_list[(TISSUE_list.BTsample > 100) &\n",
    "                          (TISSUE_list.SMTSD.values != 'Cells - Transformed fibroblasts') &\n",
    "                          (TISSUE_list.SMTSD.values != 'Cells - EBV-transformed lymphocytes')].reset_index(drop=True)\n",
    "TISSUE_list_short.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_tpm=gene_tpm.set_index(\"Name\").drop(columns=\"Description\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_feature=[0.10]\n",
    "use_target=[1.00]\n",
    "optimal_features=[5, 10,15, 20, 25, 30]\n",
    "# \"randData_cosine\"\n",
    "do_select=[\"lasso\",\"cosine\",\"randSelect\"]\n",
    "# models=[\"lsr\",\"bay\"]\n",
    "raw_df=[\"summary0\", \"summary2\", \"train0\", \"val2\", \"train_mean\",\"train_std\"]\n",
    "metric_summary=pd.DataFrame(index=TISSUE_list_short.SMTSD, \n",
    "                            columns=[\"mae_mean\",\"rmse_mean\",\"r2_mean\", \"mae_std\",\"rmse_std\",\"r2_std\"])\n",
    "model_dict = {\n",
    "    \"lsr\":  linear_model.LinearRegression(),\n",
    "    \"Lasso\": linear_model.Lasso(alpha=0.04, max_iter=1e5),\n",
    "    \"Ridge\": linear_model.Ridge(),\n",
    "    \"Bay\": linear_model.BayesianRidge()\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Apr 10 14:42:05 2020 Adipose - Subcutaneous (12121, 278) (15917, 278)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/xwj/soft/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:576: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item_labels[indexer[info_axis]]] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Apr 10 14:42:10 2020 Adipose - Subcutaneous 1 1\n",
      "Fri Apr 10 14:42:10 2020 Muscle - Skeletal (12000, 341) (14006, 341)\n",
      "Fri Apr 10 14:42:15 2020 Muscle - Skeletal 1 1\n",
      "Fri Apr 10 14:42:15 2020 Artery - Tibial (12334, 283) (15029, 283)\n",
      "Fri Apr 10 14:42:19 2020 Artery - Tibial 1 1\n",
      "Fri Apr 10 14:42:19 2020 Artery - Coronary (13079, 116) (16552, 116)\n",
      "Fri Apr 10 14:42:24 2020 Artery - Coronary 1 1\n",
      "Fri Apr 10 14:42:25 2020 Heart - Atrial Appendage (13037, 171) (15636, 171)\n",
      "Fri Apr 10 14:42:29 2020 Heart - Atrial Appendage 1 1\n",
      "Fri Apr 10 14:42:30 2020 Adipose - Visceral (Omentum) (12614, 197) (16163, 197)\n",
      "Fri Apr 10 14:42:34 2020 Adipose - Visceral (Omentum) 1 1\n",
      "Fri Apr 10 14:42:34 2020 Breast - Mammary Tissue (12602, 173) (16373, 173)\n",
      "Fri Apr 10 14:42:39 2020 Breast - Mammary Tissue 1 1\n",
      "Fri Apr 10 14:42:40 2020 Skin - Not Sun Exposed (Suprapubic) (12555, 203) (16639, 203)\n",
      "Fri Apr 10 14:42:45 2020 Skin - Not Sun Exposed (Suprapubic) 1 1\n",
      "Fri Apr 10 14:42:45 2020 Adrenal Gland (12779, 121) (16899, 121)\n",
      "Fri Apr 10 14:42:50 2020 Adrenal Gland 1 1\n",
      "Fri Apr 10 14:42:50 2020 Thyroid (12078, 256) (17121, 256)\n",
      "Fri Apr 10 14:42:56 2020 Thyroid 1 1\n",
      "Fri Apr 10 14:42:56 2020 Lung (12079, 271) (16573, 271)\n",
      "Fri Apr 10 14:43:01 2020 Lung 1 1\n",
      "Fri Apr 10 14:43:02 2020 Pancreas (12720, 156) (16277, 156)\n",
      "Fri Apr 10 14:43:07 2020 Pancreas 1 1\n",
      "Fri Apr 10 14:43:07 2020 Esophagus - Muscularis (12444, 224) (16372, 224)\n",
      "Fri Apr 10 14:43:12 2020 Esophagus - Muscularis 1 1\n",
      "Fri Apr 10 14:43:13 2020 Esophagus - Mucosa (12275, 247) (16299, 247)\n",
      "Fri Apr 10 14:43:18 2020 Esophagus - Mucosa 1 1\n",
      "Fri Apr 10 14:43:18 2020 Esophagus - Gastroesophageal Junction (13223, 139) (16665, 139)\n",
      "Fri Apr 10 14:43:23 2020 Esophagus - Gastroesophageal Junction 1 1\n",
      "Fri Apr 10 14:43:23 2020 Stomach (13096, 165) (16065, 165)\n",
      "Fri Apr 10 14:43:29 2020 Stomach 1 1\n",
      "Fri Apr 10 14:43:29 2020 Colon - Sigmoid (13164, 135) (16821, 135)\n",
      "Fri Apr 10 14:43:34 2020 Colon - Sigmoid 1 1\n",
      "Fri Apr 10 14:43:34 2020 Colon - Transverse (13111, 173) (16568, 173)\n",
      "Fri Apr 10 14:43:39 2020 Colon - Transverse 1 1\n",
      "Fri Apr 10 14:43:39 2020 Testis (12515, 163) (22094, 163)\n",
      "Fri Apr 10 14:43:46 2020 Testis 1 1\n",
      "Fri Apr 10 14:43:46 2020 Nerve - Tibial (12258, 257) (16779, 257)\n",
      "Fri Apr 10 14:43:52 2020 Nerve - Tibial 1 1\n",
      "Fri Apr 10 14:43:52 2020 Skin - Sun Exposed (Lower leg) (12059, 289) (16132, 289)\n",
      "Fri Apr 10 14:43:57 2020 Skin - Sun Exposed (Lower leg) 1 1\n",
      "Fri Apr 10 14:43:58 2020 Heart - Left Ventricle (12479, 196) (14735, 196)\n",
      "Fri Apr 10 14:44:02 2020 Heart - Left Ventricle 1 1\n",
      "Fri Apr 10 14:44:03 2020 Brain - Cerebellum (13106, 101) (18342, 101)\n",
      "Fri Apr 10 14:44:08 2020 Brain - Cerebellum 1 1\n",
      "Fri Apr 10 14:44:08 2020 Artery - Aorta (12431, 202) (16149, 202)\n",
      "Fri Apr 10 14:44:13 2020 Artery - Aorta 1 1\n",
      "Fri Apr 10 14:44:13 2020 Pituitary (13176, 104) (18507, 104)\n",
      "Fri Apr 10 14:44:19 2020 Pituitary 1 1\n",
      "Fri Apr 10 14:44:19 2020 Liver (13163, 105) (15759, 105)\n",
      "Fri Apr 10 14:44:23 2020 Liver 1 1\n"
     ]
    }
   ],
   "source": [
    "input_dict={}\n",
    "for TISSUE in TISSUE_list_short.SMTSD:\n",
    "\n",
    "    input_dict[TISSUE]={}\n",
    "    input_dict[TISSUE][\"feature\"],  input_dict[TISSUE][\"target\"] ={},{}\n",
    "    input_dict[TISSUE][\"feature\"][\"train0\"]={}\n",
    "    input_dict[TISSUE][\"feature\"][\"val2\"]={}\n",
    "    input_dict[TISSUE][\"target\"][\"train0\"]={}\n",
    "    input_dict[TISSUE][\"target\"][\"val2\"]={}\n",
    "    \n",
    "    input_dict[TISSUE][\"feature\"][\"train_mean\"]={}\n",
    "    input_dict[TISSUE][\"feature\"][\"train_std\"]={}\n",
    "    input_dict[TISSUE][\"target\"][\"train_mean\"]={}\n",
    "    input_dict[TISSUE][\"target\"][\"train_std\"]={}\n",
    "    \n",
    "    target_id = samples[(samples.SMTSD == TISSUE ) & (samples.SMAFRZE == 'RNASEQ')].loc[:,['SUBJID','SAMPID']]\n",
    "    sample_overlap = list(set(target_id.SUBJID) & set(feature_id.SUBJID))\n",
    "    feature_tpm = gene_tpm[list(feature_id[feature_id.SUBJID.isin(sample_overlap)].SAMPID)]\n",
    "    target_tpm = gene_tpm[list(target_id[target_id.SUBJID.isin(sample_overlap)].SAMPID)]\n",
    "    \n",
    "#     print(time.ctime(), TISSUE, feature_tpm.shape, target_tpm.shape)\n",
    "    feature_tpm = feature_tpm.loc[feature_tpm.all(axis=1)] #.reset_index(drop=True)\n",
    "    target_tpm = target_tpm.loc[target_tpm.all(axis=1)] #.reset_index(drop=True)\n",
    "    print(time.ctime(), TISSUE, feature_tpm.shape, target_tpm.shape)\n",
    "    \n",
    "    # split samples into 90% train+test set and 10% vaildation set\n",
    "    sample_use, sample_val2=train_test_split(sample_overlap, test_size=0.1, random_state=123, shuffle=True)\n",
    "\n",
    "    feature_tpm_use = feature_tpm[list(feature_id[feature_id.SUBJID.isin(sample_use)].SAMPID)]\n",
    "    target_tpm_use = target_tpm[list(target_id[target_id.SUBJID.isin(sample_use)].SAMPID)]\n",
    "\n",
    "    feature_tpm_val2 = feature_tpm[list(feature_id[feature_id.SUBJID.isin(sample_val2)].SAMPID)]\n",
    "    target_tpm_val2 = target_tpm[list(target_id[target_id.SUBJID.isin(sample_val2)].SAMPID)]\n",
    "    \n",
    "    # compute coefficient of variation with train set ====cv\n",
    "    for uf in use_feature:\n",
    "        percent=uf\n",
    "        \n",
    "        tissue_tpm=feature_tpm_use\n",
    "        cutoff = tissue_tpm.shape[0]*percent\n",
    "        cv=tissue_tpm.std(axis=1)/tissue_tpm.mean(axis=1)\n",
    "        mask=cv.rank(ascending=False) <= cutoff\n",
    "        tissue_tpm = tissue_tpm[mask]\n",
    "        # keep the train_mean, and train_std values before transform data\n",
    "        train_mean, train_std = tissue_tpm.mean(axis=1), tissue_tpm.std(axis=1)\n",
    "        \n",
    "        tissue_tpm = tissue_tpm.transform(lambda x: (x-x.mean())/x.std(), axis=1).round(3)       \n",
    "        input_dict[TISSUE][\"feature\"][\"train0\"][uf]=tissue_tpm\n",
    "        input_dict[TISSUE][\"feature\"][\"train_mean\"][uf]=train_mean\n",
    "        input_dict[TISSUE][\"feature\"][\"train_std\"][uf]=train_std\n",
    "        \n",
    "        tissue_tpm=feature_tpm_val2[mask]\n",
    "        for col in tissue_tpm:\n",
    "            tissue_tpm.loc[:, col]=(tissue_tpm.loc[:, col]-train_mean)/train_std\n",
    "        input_dict[TISSUE][\"feature\"][\"val2\"][uf]=tissue_tpm\n",
    "    \n",
    "    for ut in use_target: \n",
    "        percent=ut\n",
    "        \n",
    "        tissue_tpm=target_tpm_use\n",
    "        cutoff = tissue_tpm.shape[0]*percent\n",
    "        cv=tissue_tpm.std(axis=1)/tissue_tpm.mean(axis=1)\n",
    "        mask=cv.rank(ascending=False) <= cutoff\n",
    "        tissue_tpm = tissue_tpm[mask]\n",
    "        # keep the train_mean, and train_std values before transform data\n",
    "        train_mean, train_std = tissue_tpm.mean(axis=1), tissue_tpm.std(axis=1)\n",
    "       \n",
    "        tissue_tpm = tissue_tpm.transform(lambda x: (x-x.mean())/x.std(), axis=1).round(3)       \n",
    "        input_dict[TISSUE][\"target\"][\"train0\"][ut]=tissue_tpm\n",
    "        input_dict[TISSUE][\"target\"][\"train_mean\"][uf]=train_mean\n",
    "        input_dict[TISSUE][\"target\"][\"train_std\"][uf]=train_std\n",
    "\n",
    "        tissue_tpm=target_tpm_val2[mask]\n",
    "        for col in tissue_tpm:\n",
    "            tissue_tpm.loc[:, col]=(tissue_tpm.loc[:, col]-train_mean)/train_std\n",
    "        input_dict[TISSUE][\"target\"][\"val2\"][ut]=tissue_tpm\n",
    "\n",
    "    print(time.ctime(), TISSUE, len(input_dict[TISSUE][\"feature\"][\"val2\"]), len(input_dict[TISSUE][\"target\"][\"val2\"]))        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "final={}\n",
    "for uf in use_feature:\n",
    "    if uf not in final:\n",
    "        final[uf]={}\n",
    "    for ut in use_target:\n",
    "        if ut not in final[uf]:\n",
    "            final[uf][ut]={}\n",
    "        for of in optimal_features[1:2]:\n",
    "            if of not in final[uf][ut]:\n",
    "                final[uf][ut][of]={}\n",
    "            for do in do_select[1:2]:\n",
    "                if do not in final[uf][ut][of]:\n",
    "                    final[uf][ut][of][do]={}\n",
    "                # save feature names in df\n",
    "                final[uf][ut][of][do][\"selected_feature\"]={}\n",
    "                final[uf][ut][of][do][\"trained_model\"]={}\n",
    "                \n",
    "                # save pre_trained model parameters \n",
    "                for m in model_dict:\n",
    "                    if m not in final[uf][ut][of][do][\"trained_model\"]:\n",
    "                        final[uf][ut][of][do][\"trained_model\"][m]={}\n",
    "                    \n",
    "                    if m not in final[uf][ut][of][do]:\n",
    "                        final[uf][ut][of][do][m]={}\n",
    "                        for rdf in raw_df:\n",
    "                            if rdf ==\"summary0\" or rdf ==\"summary2\":\n",
    "                                final[uf][ut][of][do][m][rdf]=deepcopy(metric_summary)\n",
    "                            else:\n",
    "                                final[uf][ut][of][do][m][rdf]={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5, 10, 15, 20, 25, 30] 4 10 0.1 1.0\n"
     ]
    }
   ],
   "source": [
    "                        \n",
    "def worker_return_feature_and_final_models(p):\n",
    "    \n",
    "    y_val, y_use=Y_val[p], Y_use[p]\n",
    "    gene_error=np.zeros(9)\n",
    "\n",
    "    '''# [2] use cosine similarity way to select topN , only use TRAIN set :  a better feature selection methods??'''\n",
    "    select=X_use.columns[cos_top[p]]\n",
    "    X_use_short=X_use[select]\n",
    "        \n",
    "    '''fit model with selected feature_and_final'''\n",
    "    mymodel=model_dict[M].fit(X_use_short, y_use)\n",
    "#     y_predict = mymodel.predict(X_val_short)\n",
    "\n",
    "    return select, mymodel\n",
    "\n",
    "def worker_with_test(p):\n",
    "    \n",
    "    y_val, y_use=Y_val[p], Y_use[p]\n",
    "    gene_error=np.zeros(9)\n",
    "\n",
    "    '''# [2] use cosine similarity way to select topN , only use TRAIN set :  a better feature selection methods??'''\n",
    "    select=X_use.columns[cos_top[p]]\n",
    "    X_val_short, X_use_short=X_val[select], X_use[select]\n",
    "        \n",
    "    '''fit model with selected features'''\n",
    "    '''test error'''\n",
    "    mymodel=model_dict[M].fit(X_use_short, y_use)\n",
    "   \n",
    "    y_predict = mymodel.predict(X_val_short)\n",
    "    gene_error[6]= metrics.mean_absolute_error(y_val, y_predict)\n",
    "    gene_error[7]= np.sqrt(metrics.mean_squared_error(y_val, y_predict))\n",
    "    gene_error[8]= np.corrcoef(y_val.astype(float), y_predict)[1,0]                 \n",
    "    # get back result of train，test, validat in a long row vector\n",
    "    return gene_error\n",
    "\n",
    "'''how many CPUs to use (4 cpu ~ 30 mins)'''\n",
    "CORE=4\n",
    "\n",
    "'''set topN: best N features using train set?'''\n",
    "topN=10\n",
    "uf, ut =0.10, 1.00\n",
    "print(optimal_features, CORE, topN, uf, ut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Apr 10 14:44:24 2020 Adipose - Subcutaneous (1212, 250) (15917, 250)\n",
      "Fri Apr 10 14:44:24 2020 Adipose - Subcutaneous cosine 10 Bay\n",
      "Fri Apr 10 14:46:01 2020 Muscle - Skeletal (1200, 306) (14006, 306)\n",
      "Fri Apr 10 14:46:01 2020 Muscle - Skeletal cosine 10 Bay\n"
     ]
    }
   ],
   "source": [
    "for TISSUE in list(input_dict.keys())[:2]:\n",
    "# for TISSUE in input_dict:\n",
    "\n",
    "    feature_tpm, target_tpm = input_dict[TISSUE][\"feature\"][\"train0\"][uf], input_dict[TISSUE][\"target\"][\"train0\"][ut]\n",
    "    feature_tpm_val, target_tpm_val = input_dict[TISSUE][\"feature\"][\"val2\"][uf], input_dict[TISSUE][\"target\"][\"val2\"][ut]\n",
    "    \n",
    "#     feature_tpm, target_tpm = input_dict_rand[TISSUE][\"feature\"], input_dict_rand[TISSUE][\"target\"]\n",
    "    print(time.ctime(), TISSUE, feature_tpm.shape, target_tpm.shape)\n",
    "    X_use, Y_use, X_val, Y_val =feature_tpm.transpose(), target_tpm.transpose(), feature_tpm_val.transpose(), target_tpm_val.transpose()\n",
    "    \n",
    "    for M in list(model_dict.keys())[3:]:\n",
    "        # take evaluation results df (valid OR test error ) and df0 (train error)\n",
    "        for ds in do_select[1:2]: #do_select=[\"lasso\",\"cosine\",\"randSelect\"]        \n",
    "            \n",
    "            for topN in optimal_features[1:2]:\n",
    "                print(time.ctime(), TISSUE, ds, topN, M)\n",
    "\n",
    "                cos=pd.DataFrame(data=metrics.pairwise.cosine_similarity(X=X_use.transpose(), Y=Y_use.transpose(), dense_output=True))\n",
    "                cos_top=(abs(cos).rank(axis=0, ascending=False) <= topN)\n",
    "                cos_top.columns = target_tpm.index\n",
    "                \n",
    "                final[uf][ut][topN][ds][\"selected_feature\"][TISSUE]=pd.DataFrame(index=Y_use.columns, columns=range(topN))\n",
    "                final[uf][ut][topN][ds][\"trained_model\"][M][TISSUE]={}\n",
    "                \n",
    "                '''save final features and model parameters'''\n",
    "                for p in Y_use.columns:\n",
    "                    # select, mymodel.best_estimator_ \n",
    "                    final[uf][ut][topN][ds][\"selected_feature\"][TISSUE].loc[p,:], \\\n",
    "                    final[uf][ut][topN][ds][\"trained_model\"][M][TISSUE][p] \\\n",
    "                    =worker_return_feature_and_final_models(p) \n",
    "                \n",
    "                '''save test error'''\n",
    "                with Pool(CORE) as cpu_pool:\n",
    "                    gene_error_in_rows = cpu_pool.map(partial(worker_with_test), Y_use.columns)\n",
    "\n",
    "                gene_error=pd.DataFrame(data=gene_error_in_rows, index=Y_use.columns)                  \n",
    "                metric_df2=pd.DataFrame(index=Y_use.columns, columns=[\"mae\",\"rmse\",\"r2\"])\n",
    "                metric_df2[\"mae\"], metric_df2[\"rmse\"], metric_df2[\"r2\"]=gene_error[6],gene_error[7], gene_error[8]\n",
    "                final[uf][ut][topN][ds][M][\"val2\"][TISSUE]=metric_df2\n",
    "                final[uf][ut][topN][ds][M][\"summary2\"].loc[TISSUE,:]=list(metric_df2.mean(axis=0))+ list(metric_df2.std(axis=0)) \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 / 15917\n",
      "1000 / 15917\n",
      "2000 / 15917\n",
      "3000 / 15917\n",
      "4000 / 15917\n",
      "5000 / 15917\n",
      "6000 / 15917\n",
      "7000 / 15917\n",
      "8000 / 15917\n",
      "9000 / 15917\n",
      "10000 / 15917\n",
      "11000 / 15917\n",
      "12000 / 15917\n",
      "13000 / 15917\n",
      "14000 / 15917\n",
      "15000 / 15917\n"
     ]
    }
   ],
   "source": [
    "# BGEX predict with pretrained models\n",
    "TISSUE=\"Adipose - Subcutaneous\"\n",
    "ds=\"cosine\"\n",
    "topN, uf, ut =10, 0.10, 1.00\n",
    "M=\"Bay\" # BayesianRidge regression\n",
    "\n",
    "feature_std=input_dict[TISSUE][\"feature\"][\"train_std\"][uf]\n",
    "feature_mean=input_dict[TISSUE][\"feature\"][\"train_mean\"][uf]\n",
    "\n",
    "target_std=input_dict[TISSUE][\"target\"][\"train_std\"][uf]\n",
    "target_mean=input_dict[TISSUE][\"target\"][\"train_mean\"][uf]\n",
    "\n",
    "# input a blood gene expression profile in TPM value.\n",
    "# Here use TWO training sample as tutorial example \n",
    "# step 1. take releveant features from the raw blood profile\n",
    "raw_blood_profile_in_TPM=gene_tpm_raw.loc[:, [\"Name\",\"GTEX-111YS-0006-SM-5NQBE\", \"GTEX-1128S-0005-SM-5P9HI\"]].set_index(\"Name\")\n",
    "new_blood_profile = raw_blood_profile_in_TPM.reindex(feature_mean.index)\n",
    "\n",
    "# step 2. normalize by predefined mean and std\n",
    "for col in new_blood_profile:\n",
    "    new_blood_profile.loc[:, col] = (np.log(new_blood_profile.loc[:, col] +1) - feature_mean)/feature_std\n",
    "\n",
    "# step 3. use pre-trained model to predict gene expression of new sample's tissue\n",
    "predicted_tissue_profile=pd.DataFrame(index=target_mean.index, columns=new_blood_profile.columns) \n",
    "\n",
    "total=len(target_mean)\n",
    "# for p in target_mean.index:\n",
    "for i in range(total):\n",
    "    p=target_mean.index[i]\n",
    "    # select, mymodel.best_estimator_ \n",
    "    selected=final[uf][ut][topN][ds][\"selected_feature\"][TISSUE].loc[p,:]\n",
    "    mymodel=final[uf][ut][topN][ds][\"trained_model\"][M][TISSUE][p]\n",
    "    X_val_short=new_blood_profile.transpose()[selected]\n",
    "    predicted_tissue_profile.loc[p,:] = mymodel.predict(X_val_short) \n",
    "    if (i % 1000==0) :\n",
    "        print(i,\"/\", total)\n",
    "    \n",
    "# step 4. revert predicted value to TPM\n",
    "for col in predicted_tissue_profile:\n",
    "    predicted_tissue_profile.loc[:, col] = np.e**(predicted_tissue_profile.loc[:, col] * target_std + target_mean) -1\n",
    "\n",
    "# step 5. filter by empirical predictability mark\n",
    "thres_mae, thres_r2 = 0.7, 0.3\n",
    "d0=final[uf][ut][topN][ds][M][\"val2\"][TISSUE]\n",
    "predictable_gene=(d0[\"mae\"] < thres_mae) & (d0[\"r2\"] > thres_r2) \n",
    "predicted_tissue_profile_short=predicted_tissue_profile[predictable_gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
